onchain/cairo/src/staking/staking.cairo

```diff
@@ -0,0 +1,312 @@
+            ((self.balance_of.entry(account).read()
+                * (self.reward_per_token() - self.user_reward_per_token_paid.entry(account).read()))
+                / ONE_E18)
+                + self.rewards.entry(account).read()
+        }
+
+        fn staking_token(self: @ComponentState<TContractState>) -> ContractAddress {
+            self.staking_token.read()
+        }
+
+        fn rewards_token(self: @ComponentState<TContractState>) -> ContractAddress {
+            self.rewards_token.read()
+        }
+
+        fn duration(self: @ComponentState<TContractState>) -> u256 {
+            self.duration.read()
+        }
+
+        fn finish_at(self: @ComponentState<TContractState>) -> u256 {
+            self.finish_at.read()
+        }
+
+        fn updated_at(self: @ComponentState<TContractState>) -> u256 {
+            self.updated_at.read()
+        }
+
+        fn reward_rate(self: @ComponentState<TContractState>) -> u256 {
+            self.reward_rate.read()
+        }
+
+        fn reward_per_token_stored(self: @ComponentState<TContractState>) -> u256 {
+            self.reward_per_token_stored.read()
+        }
+
+        fn user_reward_per_token_paid(
+            self: @ComponentState<TContractState>, user: ContractAddress
+        ) -> u256 {
+            self.user_reward_per_token_paid.entry(user).read()
+        }
+
+        fn rewards(self: @ComponentState<TContractState>, user: ContractAddress) -> u256 {
+            self.rewards.entry(user).read()
+        }
+
+        fn total_supply(self: @ComponentState<TContractState>) -> u256 {
+            self.total_supply.read()
+        }
+
+        fn balance_of(self: @ComponentState<TContractState>, user: ContractAddress) -> u256 {
+            self.balance_of.entry(user).read()
+        }
+
+        fn owner(self: @ComponentState<TContractState>) -> ContractAddress {
+            self.owner.read()
+        }
+
+        fn return_block_timestamp(self: @ComponentState<TContractState>) -> u256 {
+            get_block_timestamp().try_into().unwrap()
+        }
+    }
+
+    #[generate_trait]
+    pub impl InternalImpl<
+        TContractState, +HasComponent<TContractState>
+    > of InternalTrait<TContractState> {
+        /// Initializes the contract by setting the owner, staking_token and reward_token.
+        /// To prevent reinitialization, this should only be used inside of a contract's
+        /// constructor.
+        fn initializer(
+            ref self: ComponentState<TContractState>,
+            owner: ContractAddress,
+            staking_token: ContractAddress,
+            reward_token: ContractAddress
+        ) {
+            self.owner.write(owner);
+            self.staking_token.write(staking_token);
+            self.rewards_token.write(reward_token);
+        }
+
+        fn update_reward(ref self: ComponentState<TContractState>, account: ContractAddress) {
```

Comment:
better to have _ as convention for internal functions please

Reply:
> better to have _ as convention for internal functions please

Fixed